nearcore/src/state_sync.rs

```diff
@@ -194,79 +233,110 @@ async fn state_sync_dump(
-                                part_id,
-                                num_parts,
-                            );
-                            if let Err(err) =
-                                external.put_state_part(&state_part, shard_id, &location).await
-                            {
-                                res = Some(Error::Other(err.to_string()));
-                                break;
+                        let missing_parts = get_missing_state_parts_for_epoch(
+                            shard_id,
+                            &chain_id,
+                            &epoch_id,
+                            epoch_height,
+                            num_parts,
+                            &external,
+                        )
+                        .await;
+
+                        match missing_parts {
+                            Err(err) => {
+                                tracing::debug!(target: "state_sync_dump", shard_id, ?err, "get_missing_state_parts_for_epoch error");
+                                Err(Error::Other(format!(
+                                    "get_missing_state_parts_for_epoch failed"
+                                )))
                             }
-                            update_progress(
-                                &shard_id,
-                                &epoch_id,
-                                epoch_height,
-                                &sync_hash,
-                                part_id,
-                                num_parts,
-                                state_part.len(),
-                                &chain,
-                            );
-
-                            // Stop if the node is stopped.
-                            // Note that without this check the state dumping thread is unstoppable, i.e. non-interruptable.
-                            if !keep_running.load(std::sync::atomic::Ordering::Relaxed) {
-                                res = Some(Error::Other("Stopped".to_owned()));
-                                break;
+                            Ok(parts_not_dumped) if parts_not_dumped.is_empty() => {
+                                Ok(Some(StateSyncDumpProgress::AllDumped {
+                                    epoch_id,
+                                    epoch_height,
+                                    num_parts: Some(num_parts),
+                                }))
+                            }
+                            Ok(parts_not_dumped) => {
+                                let mut parts_to_dump = parts_not_dumped.clone();
+                                let cnt_parts_to_dump: u64 =
+                                    parts_to_dump.len().try_into().unwrap();
+                                let mut cnt_parts_dumped = num_parts - cnt_parts_to_dump;
+                                let timer = Instant::now();
+                                while timer.elapsed().as_secs() <= 60 && !parts_to_dump.is_empty() {
```

Comment:
Please respect `keep_running`, otherwise a node can't be stopped with Ctrl+C.

Reply:
I'll move the stop down below to the clause in while